<a href="https://colab.research.google.com/github/sherlyekawindiani/data-science-2026/blob/master/Pertemuan2_Sherly_Eka_Windiani_240401010222.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## PERTEMUAN 2: Struktur Data Python, NumPy & Pandas
## Nama : Sherly Eka Windiani
## NIM  : 240401010222

# A. Latihan CODE di Modul

## 1. Import Library

In [ ]:
import pandas as pd
import numpy as np

## 2. Struktur Data Python


### 2.1 List



In [ ]:
# Membuat list
data = [10, 20, 'Python', True, 3.14]

# Akses & slicing
print(data[0]) # → 10
print(data[-1]) # → 3.14 (dari belakang)
print(data[1:3]) # → [20, 'Python']

# Manipulasi
data.append(99) # tambah di akhir
data.insert(1, 55) # sisipkan di posisi 1
data.remove(20) # hapus nilai pertama = 20
# data.sort() # urutkan (jika tipe sama) - This line causes a TypeError due to mixed data types

10
3.14
[20, 'Python']


###  2.2 Tuple

Tuple mirip dengan list, namun bersifat immutable. Artinya sekali dibuat, isinya tidak
dapat diubah.
**Kapan Menggunakan Tuple?**
Gunakan tuple ketika data bersifat tetap (immutable), seperti koordinat (lat, lon), tanggal
(year, month, day), atau konfigurasi sistem. Gunakan list ketika data perlu dimodifikasi
secara dinamis.

### 2.3 Dictionary

In [ ]:
#Python — Dictionary Operations

mahasiswa = {
'nama': 'Budi Santoso',
'nim': '2024001',
'prodi': 'Informatika',
'ipk': 3.85
}

# Akses nilai
print(mahasiswa['nama']) # → 'Budi Santoso'
print(mahasiswa.get('usia', 0)) # default jika key tak ada

# Update dan hapus
mahasiswa['angkatan'] = 2024 # tambah key baru
del mahasiswa['nim'] # hapus key

Budi Santoso
0


### 2.4 Set

Set adalah koleksi tak berurutan (unordered) yang hanya menyimpan nilai unik —
duplikat otomatis dihapus.

## 3. NumPy: Komputasi Numerik

### 3.1 Membuat NumPy Array

In [ ]:
import numpy as np
# Dari list Python
a1D = np.array([1, 2, 3, 4, 5]) # 1D
a2D = np.array([[1,2,3],[4,5,6]]) # 2D

# Array dengan nilai inisial
zeros = np.zeros((3, 4)) # nol semua
ones = np.ones((2, 3)) # satu semua
eye = np.eye(4) # matriks identitas
full = np.full((2, 2), 7) # nilai konstan

# Sequence
arange = np.arange(0, 20, 2) # 0,2,4,...18
linsp = np.linspace(0, 1, 11) # 11 titik merata

# Array acak (reproducible)
rng = np.random.default_rng(seed=42)
rand_int = rng.integers(0, 100, (3, 4))
rand_flt = rng.normal(loc=0, scale=1, size=(100,))


### 3.2 Operasi Vektorisasi dan Fungsi Statistik

In [ ]:
# Operasi aritmetika vektorisasi
a = np.array([1, 2, 3, 4, 5])
b = np.array([10, 20, 30, 40, 50])
print(a + b) # → [11 22 33 44 55]
print(a * b) # → [10 40 90 160 250]
print(a ** 2) # → [1 4 9 16 25]

# Universal functions (ufuncs)
np.sqrt(a) # akar kuadrat tiap elemen
np.exp(a) # eksponensial (e^x)
np.log(b) # logaritma natural

# Fungsi statistik
data = rng.normal(0, 1, 1000)
print(f'Mean: {np.mean(data):.4f}')
print(f'Std: {np.std(data):.4f}')
print(f'Min: {np.min(data):.4f}')
print(f'Max: {np.max(data):.4f}')
print(f'Q1: {np.percentile(data,25):.4f}')
print(f'Q3: {np.percentile(data,75):.4f}')

[11 22 33 44 55]
[ 10  40  90 160 250]
[ 1  4  9 16 25]
Mean: -0.0163
Std: 1.0008
Min: -3.6484
Max: 3.1789
Q1: -0.6897
Q3: 0.6163


## 4. Pandas: Manipulasi DataFrame

### 4.1 Pandas Series


In [ ]:
import pandas as pd

# Series dari list — index default 0,1,2...
s1 = pd.Series([85, 92, 78, 95, 88])

# Series dengan index kustom
nilai = pd.Series(
[85, 92, 78, 95],
index=['Alice','Bob','Caca','Dina']
)
print(nilai['Bob']) # → 92
print(nilai[nilai>80]) # filter
print(nilai.mean()) # → 87.5
print(nilai.describe()) # statistik lengkap

92
Alice    85
Bob      92
Dina     95
dtype: int64
87.5
count     4.000000
mean     87.500000
std       7.593857
min      78.000000
25%      83.250000
50%      88.500000
75%      92.750000
max      95.000000
dtype: float64


### 4.2 Pandas DataFrame

In [ ]:
# Membuat DataFrame dari dictionary
data = {
'nama': ['Ali','Budi','Caca','Dina'],
'usia': [22, 25, 21, 23],
'kota': ['Jakarta','Bandung','Surabaya','Medan'],
'ipk': [3.8, 3.5, 3.9, 3.7]
}
df = pd.DataFrame(data)
# Informasi dasar
print(df.shape) # → (4, 4)
print(df.dtypes) # tipe data per kolom
print(df.columns) # nama-nama kolom

(4, 4)
nama     object
usia      int64
kota     object
ipk     float64
dtype: object
Index(['nama', 'usia', 'kota', 'ipk'], dtype='object')


### 4.3 Memuat Data CSV

In [ ]:
# Memuat dari file lokal
#df = pd.read_csv('data/penjualan.csv')

# Memuat dari URL (dataset publik)
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# Parameter penting Untuk File Lokal
# df = pd.read_csv('data.csv',
# sep=';', # pemisah kolom (default ',')
# encoding='utf-8', # encoding file
# index_col='ID', # kolom sebagai index
# parse_dates=['tgl'],# parse kolom tanggal
# na_values=['N/A','none'] # nilai kosong
# )

# Eksplorasi standar setelah load
df.head() # 5 baris pertama
df.info() # tipe data & missing
df.describe() # statistik deskriptif


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### 4.4 Filtering (Boolean Indexing)

In [ ]:
# Python — Boolean Indexing

# Contoh: Dataset Titanic
df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")

# Filter satu kondisi
penumpang_selamat = df[df['Survived'] == 1]
penumpang_wanita = df[df['Sex'] == 'female']

# Filter kondisi ganda (& = AND, | = OR)
wanita_kelas1 = df[(df['Sex']=='female') &
(df['Pclass']==1)]

# Seleksi kolom tertentu
df[['Name','Age','Sex','Survived']]

,Name,Age,Sex,Survived
0,"Braund, Mr. Owen Harris",22.0,male,0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,female,1
2,"Heikkinen, Miss. Laina",26.0,female,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,female,1
4,"Allen, Mr. William Henry",35.0,male,0
...,...,...,...,...
886,"Montvila, Rev. Juozas",27.0,male,0
887,"Graham, Miss. Margaret Edith",19.0,female,1
888,"Johnston, Miss. Catherine Helen ""Carrie""",NaN,female,0
889,"Behr, Mr. Karl Howell",26.0,male,1


### 4.5 Groupby dan Agregasi

In [ ]:
# Python — Groupby & Agregasi

# Pola dasar: groupby → aggregation
df.groupby('Pclass')['Survived'].mean()

# Agregasi multiple kolom & fungsi
hasil = df.groupby('Pclass').agg(
total = ('PassengerId', 'count'),
selamat = ('Survived', 'sum'),
pct_selamat= ('Survived', 'mean'),
rata_usia = ('Age', 'mean'),
max_fare = ('Fare', 'max')
)
hasil = hasil.round(3)
print(hasil)

# GroupBy multi-kolom
df.groupby(['Pclass','Sex'])['Survived']\
.mean().unstack().round(3)

        total  selamat  pct_selamat  rata_usia  max_fare
Pclass                                                  
1         216      136        0.630     38.233   512.329
2         184       87        0.473     29.878    73.500
3         491      119        0.242     25.141    69.550


Sex,female,male
Pclass,,
1,0.968,0.369
2,0.921,0.157
3,0.500,0.135


# B. Praktikum Dataset Titanic

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

df = pd.read_csv(url)

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [ ]:
print(df.shape)

(891, 12)


In [ ]:
print(df.columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


# C. Menjawab Pertanyaan Latihan Q1 - Q6

## Q1. Berapa jumlah baris dan kolom pada dataset Titanic? Kolom apa saja yang tersedia?

In [ ]:
print("Jumlah baris dan kolom:")
print(df.shape)

print("\nNama Kolom:")
print(df.columns)

Jumlah baris dan kolom:
(891, 12)

Nama Kolom:
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


### Interpretasi Q1

Berdasarkan hasil analisis, dataset Titanic terdiri dari  **891 baris dan 12 kolom.** Jumlah baris menunjukkan bahwa terdapat 891 data penumpang yang tercatat dalam dataset, sedangkan 12 kolom menunjukkan atribut atau informasi yang dimiliki oleh setiap penumpang.

Kolom yang tersedia dalam dataset yaitu PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, dan Embarked. Kolom-kolom tersebut berisi informasi mengenai identitas penumpang, status keselamatan, kelas penumpang, nama, jenis kelamin, usia, jumlah saudara atau pasangan yang ikut, jumlah orang tua atau anak yang ikut, nomor tiket, tarif tiket, nomor kabin, serta pelabuhan keberangkatan.

## Q2 Kolom mana saja yang memiliki missing values? Berapa jumlah dan proporsinya (%) per kolom?

In [ ]:
# Menghitung jumlah missing value per kolom
missing = df.isnull().sum()

# Menghitung persentase missing value per kolom
persentase = (missing / len(df)) * 100

print("Jumlah Missing Value")
print(missing)

print("\nPersentase Missing Value")
print(persentase)

Jumlah Missing Value
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Persentase Missing Value
PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64


### Interpretasi Q2

**Kolom Cabin memiliki jumlah missing value tertinggi, yaitu 687 data** atau 77,10% dari keseluruhan dataset. Sementara itu, kolom Age memiliki 177 data kosong (19,87%) dan Embarked hanya 2 data kosong (0,22%).

Banyaknya data yang hilang pada kolom Cabin dan Age menunjukkan bahwa kualitas data pada kedua kolom tersebut perlu diperhatikan karena dapat memengaruhi hasil analisis yang menggunakan informasi kabin dan usia penumpang.

## Q3 Berapa persen penumpang yang selamat (Survived=1)? Gunakan operasi NumPy atau Pandas.

In [ ]:
persen_selamat = df["Survived"].mean() * 100

print(f"Persentase penumpang yang selamat: {persen_selamat:.2f}%")

Persentase penumpang yang selamat: 38.38%


### Interpretasi Q3

**Persentase penumpang yang selamat adalah 38,38%,** yang berarti kurang dari setengah penumpang berhasil bertahan hidup dalam tragedi Titanic.

Hal ini menunjukkan bahwa tingkat keselamatan penumpang relatif rendah, dengan mayoritas penumpang dalam dataset tercatat tidak selamat.

## Q4 Filter penumpang wanita dari kelas 1 (Pclass=1). Berapa jumlahnya dan berapa rata-rata usianya?

In [ ]:
# Filter wanita kelas 1
wanita_kelas1 = df[
    (df["Sex"] == "female") &
    (df["Pclass"] == 1)
]

# Jumlah wanita kelas 1
jumlah_wanita = len(wanita_kelas1)

# Rata-rata usia
rata_usia = wanita_kelas1["Age"].mean()

print("Jumlah wanita kelas 1:", jumlah_wanita)
print("Rata-rata usia:", round(rata_usia, 2))

Jumlah wanita kelas 1: 94
Rata-rata usia: 34.61


### Interpretasi Q4

**Terdapat 94 penumpang wanita yang berada pada kelas 1 dengan rata-rata usia 34,61 tahun**.

Hasil ini menunjukkan bahwa kelompok wanita kelas 1 didominasi oleh penumpang usia dewasa. Informasi ini dapat digunakan untuk analisis lebih lanjut mengenai hubungan antara usia, kelas penumpang, dan tingkat keselamatan.

## Q5 Gunakan groupby untuk menghitung tingkat keselamatan (mean Survived) per kelas penumpang. Kelas mana yang paling tinggi?

In [ ]:
# Mengelompokkan data berdasarkan kelas penumpang (Pclass)
# lalu menghitung rata-rata nilai Survived pada setiap kelas
keselamatan = df.groupby("Pclass")["Survived"].mean() * 100

print("Tingkat keselamatan berdasarkan kelas penumpang (%):")
print(keselamatan.round(2))

Tingkat keselamatan berdasarkan kelas penumpang (%):
Pclass
1    62.96
2    47.28
3    24.24
Name: Survived, dtype: float64


### Interpretasi Q5

**Tingkat keselamatan tertinggi terdapat pada penumpang kelas 1, yaitu sebesar 62,96%,** diikuti kelas 2 sebesar 47,28%, dan kelas 3 sebesar 24,24%.

Perbedaan ini menunjukkan bahwa penumpang dari kelas yang lebih tinggi memiliki peluang selamat yang lebih besar dibandingkan penumpang kelas bawah. Hal ini mengindikasikan bahwa kelas penumpang kemungkinan memiliki pengaruh terhadap tingkat keselamatan pada tragedi Titanic.

## Q6 Buat analisis tambahan berdasarkan variabel lain yang Anda anggap menarik. Sertakan interpretasi singkat.

In [ ]:
# Menghitung jumlah penumpang pada setiap kelas

jumlah_penumpang = df["Pclass"].value_counts().sort_index()

print("Jumlah penumpang per kelas:")
print(jumlah_penumpang)

Jumlah penumpang per kelas:
Pclass
1    216
2    184
3    491
Name: count, dtype: int64


### Interpretasi Q6

Berdasarkan hasil analisis, **Kelas 3 merupakan kelas dengan jumlah penumpang terbanyak**, yaitu 491 orang atau lebih dari setengah total penumpang dalam dataset.

Sementara itu, kelas 1 dan kelas 2 memiliki jumlah penumpang yang jauh lebih sedikit, masing-masing 216 dan 184 orang. Hal ini menunjukkan bahwa mayoritas penumpang Titanic berasal dari kelas ekonomi (kelas 3).